transports/janus_mqtt.c

```diff
@@ -0,0 +1,648 @@
+		if(rc != MQTTASYNC_SUCCESS) {
+			JANUS_LOG(LOG_ERR, "Can't subscribe to MQTT topic: %s, return code: %d\n", ctx->subscribe.topic, rc);
+		}
+	}
+	else if(janus_mqtt_admin_api_enabled_) {
+		int rc = janus_mqtt_client_admin_subscribe(context);
+		if(rc != MQTTASYNC_SUCCESS) {
+			JANUS_LOG(LOG_ERR, "Can't subscribe to MQTT admin topic: %s, return code: %d\n", ctx->admin.subscribe.topic, rc);
+		}
+	}
+}
+
+void janus_mqtt_client_connect_failure(void *context, MQTTAsync_failureData *response) {
+	int rc = response ? response->code : 0;
+	JANUS_LOG(LOG_ERR, "MQTT client has been failed connecting to the broker, return code: %d. Reconnecting...\n", rc);
+	// automatic reconnect
+}
+
+int janus_mqtt_client_reconnect(janus_mqtt_context *ctx) {
+	MQTTAsync_disconnectOptions options = MQTTAsync_disconnectOptions_initializer;
+	options.onSuccess = janus_mqtt_client_reconnect_success;
+	options.onFailure = janus_mqtt_client_reconnect_failure;
+	options.context = ctx;
+	options.timeout = ctx->disconnect.timeout;
+	return MQTTAsync_disconnect(ctx->client, &options);
+}
+
+void janus_mqtt_client_reconnect_success(void *context, MQTTAsync_successData *response) {
+	JANUS_LOG(LOG_INFO, "MQTT client has been successfully disconnected. Reconnecting...\n");
+
+	int rc = janus_mqtt_client_connect(context);
+	if(rc != MQTTASYNC_SUCCESS) {
+		JANUS_LOG(LOG_FATAL, "Can't connect to MQTT broker, return code: %d\n", rc);
+	}
+}
+
+void janus_mqtt_client_reconnect_failure(void *context, MQTTAsync_failureData *response) {
+	int rc = response ? response->code : 0;
+	JANUS_LOG(LOG_ERR, "MQTT client has been failed reconnecting from MQTT broker, return code: %d\n", rc);
+}
+
+int janus_mqtt_client_disconnect(janus_mqtt_context *ctx) {
+	MQTTAsync_disconnectOptions options = MQTTAsync_disconnectOptions_initializer;
+	options.onSuccess = janus_mqtt_client_disconnect_success;
+	options.onFailure = janus_mqtt_client_disconnect_failure;
+	options.context = ctx;
+	options.timeout = ctx->disconnect.timeout;
+	return MQTTAsync_disconnect(ctx->client, &options);
+}
+
+void janus_mqtt_client_disconnect_success(void *context, MQTTAsync_successData *response) {
+	JANUS_LOG(LOG_INFO, "MQTT client has been successfully disconnected. Destroying the client...\n");
+
+	janus_mqtt_client_destroy_context(context);
+}
+
+void janus_mqtt_client_disconnect_failure(void *context, MQTTAsync_failureData *response) {
+	int rc = response ? response->code : 0;
+	JANUS_LOG(LOG_ERR, "Can't disconnect from MQTT broker, return code: %d\n", rc);
+
+	janus_mqtt_client_destroy_context(context);
+}
+
+int janus_mqtt_client_subscribe(janus_mqtt_context *ctx) {
+	MQTTAsync_responseOptions options = MQTTAsync_responseOptions_initializer;
+	options.onSuccess = janus_mqtt_client_subscribe_success;
+	options.onFailure = janus_mqtt_client_subscribe_failure;
+	options.context = ctx;
+	return MQTTAsync_subscribe(ctx->client, ctx->subscribe.topic, ctx->subscribe.qos, &options);
```

Comment:
But the way it works now, aren't you trying to first subscribe to the Janus API, and only in the success callback, subscribe to the Admin API after that? If the Janus API is disabled, wouldn't the first subscription fail? Not really familiar with the library, so please forgive me if the question is silly :smile: